### PR TITLE
Add Persistent Sampling Algorithm to BlackJAX

### PR DESCRIPTION
**Pull request by ChrisBoettner**
This PR adds the persistent sampling SMC algorithm by Karamanis et al. 2025 to BlackJAX. This was first suggested in https://github.com/blackjax-devs/blackjax/issues/786.

## Overview

Persistent sampling is a family of SMC algorithms introduced by Karamanis et al. (2025) that aims to improve the efficiency of standard (tempered) SMC. Unlike traditional SMC which discards particles after resampling, persistent sampling maintains an ensemble of historical particles from all previous iterations for for improved stability and more reliable marginal likelihood estimates.

**Advantages**:
- Historical particles provide a larger effective sample size
- Lower variance estimates for marginal likelihood $Z$
- Lower variance when calculating expectations
- Adaptive effective sample size which can exceed 1.0, unlike standard SMC

**Trade-offs**:
- Introduces small bias (proven to be consistent)
- Higher memory footprint (stores all particles)
- More arithmetic operations per iteration

The implementation includes a fixed-tempering schedule and adaptive schedule variant. The implementation closely follows that of the `tempered` and `adaptive_tempered` SMC algorithms, and aims to be a plug-in replacement, if desired.

## Fixed-Schedule Persistent Sampling

The core persistent sampling implementation follows Algorithm 2 from Karamanis et al. (2025), with the change that the tempering schedule has be to provided rather than being determined adaptively. The implementation closely follows `tempered_smc`.

Since all particles are tracked (rather than just the particles in the current iteration), we have to pass an initial `n_schedule` parameter in order to pre-allocate all necessary arrays in the `PSState` (see below), which get filled with each iteration.

## Adaptive Persistent Sampling

This version follows Algorithm 2 directly. The implementation closely follows `adaptive_tempered_smc`

Since the number of iterations is not known beforehand, a `max_iterations` has to be passed. All state arrays are pre-allocated to this maximum size. The inference loop should be written in such a way that it breaks one `max_iterations` is reached. (This is a design choice I was not quite sure about.) Unlike standard SMC, the Persistent Sampling algorithm reach an effective sample size > 1. In principle, the inference loop can be run even after reaching $\lambda=1$, until the desired ESS is reached.

## API Design

Both samplers follow BlackJAX conventions and expose consistent APIs:

```python
import blackjax

# Fixed schedule persistent sampling
ps = blackjax.persistent_sampling_smc(
    logprior_fn: Callable,
    loglikelihood_fn: Callable,
    n_schedule: int | Array, # must match the number of steps in the tempering schedule
    mcmc_step_fn: Callable,
    mcmc_init_fn: Callable,
    mcmc_parameters: dict,
    resampling_fn: Callable,
    num_mcmc_steps: int = 10,
    update_strategy: Callable = update_and_take_last,
)

# Adaptive persistent sampling
aps = blackjax.adaptive_persistent_sampling_smc(
    logprior_fn: Callable,
    loglikelihood_fn: Callable,
    max_iterations: int | Array, # required max number of iterations
    mcmc_step_fn: Callable,
    mcmc_init_fn: Callable,
    mcmc_parameters: dict,
    resampling_fn: Callable,
    target_ess: float | Array = 3, # can exceed 1
    num_mcmc_steps: int = 10,
    update_strategy: Callable = update_and_take_last,
    root_solver: Callable = solver.dichotomy,
)
```

Both samplers:
- Support PyTree inputs for structured parameters
- Return `PSState` with full persistent ensemble
- Provide iteration information via `PSInfo`
- Are fully compatible with JAX transformations (jit, vmap)
- Should work with any BlackJAX MCMC kernel (HMC, NUTS, MALA, etc.)
- Support all standard resampling methods (systematic, multinomial, etc.)

## State Structure

The `PSState` NamedTuple contains:
- `persistent_particles`: All particles from all iterations (padded)
- `persistent_log_likelihoods`: Log-likelihoods for persistent ensemble
- `persistent_log_Z`: History of normalizing constants
- `tempering_schedule`: History of λ values
- `iteration`: Current iteration index

Convenience properties:
- `particles`: Current iteration particles
- `lmbda`: Current tempering parameter
- `log_Z`: Latest log normalizing constant estimate
- `persistent_weights`: Current importance weights
- `num_particles`: Number of particles per iteration

Further, a convience function `remove_padding` is implemented that takes a `PSState` and removes the zero-padding based on the current iteration.

## Implementation

The implementation is organized into two main files:

- `blackjax/smc/persistent_sampling.py`: Core persistent sampling implementation (~800 lines)
  - State management with padded arrays for JIT compatibility
  - Weight computation following paper equations
  - Resampling from persistent ensemble
  - Fixed tempering schedule kernel
  - Kernel matched `tempered_SMC` kernel signature 
    - (rng_key, state, num_mcmc_steps, lmbda, mcmc_parameters,) -> (new_state, info)

- `blackjax/smc/adaptive_persistent_sampling.py`: Adaptive schedule variant (~250 lines)
  - Automatic λ calculation using implemented solvers, by default dichotomy solver
  - Target ESS-based adaptation
  - Inference loops can have stopping condition based on iteration, $\lambda$ and ESS
  - Kernel matched `adaptive_tempered_SMC` kernel signature
    - (rng_key, state, num_mcmc_steps, mcmc_parameters,) -> (new_state, info)

Both implementations are fully JIT-compilable. The padded array approach trades memory for performance.

## Testing

Comprehensive test coverage with 15+ tests across multiple test classes:

**PersistentSamplingUnitTest** (8 tests):
- Initialization with various PyTree structures (arrays, dicts, nested dicts)
- Padding removal functionality
- Log normalizing constant computation (Equation 16)
- Persistent weight calculation (Equations 14-15)
- Persistent ESS calculation with uniform and single weights
- Resampling from persistent ensemble
- Complex PyTree resampling scenarios

**PersistentSamplingStateUpdateTest** (1 test):
- State update mechanics across iterations

**PersistentSamplingPosteriorTest** (2 tests, directly adapted from `test_tempered_smc`):
- Fixed schedule sampling on linear regression problem
- Adaptive schedule sampling on linear regression problem

**NormalizingConstantTest** (2 tests, directly adapted from `test_tempered_smc`):
- Marginal likelihood estimation with fixed schedule
- Marginal likelihood estimation with adaptive schedule
- Validation against analytical solution for multivariate normal

All tests pass with full coverage of core functionality, including JIT compilation and complex PyTree support.

## Integration

The new samplers are integrated into the package structure:
- Added to `blackjax.__init__.py` as top-level APIs
- Included in `smc_family` alongside existing SMC methods
- Exported from `blackjax.smc` module
- Follow existing naming conventions and patterns

## Documentation and Example

I have added comprehensive comments and Doc strings to the code that should cover all details and design choices.
I have also added an example use as a PR to the Sampling book, see https://github.com/blackjax-devs/sampling-book/pull/72

## References

Karamanis, M. & Seljak, U. (2025). "Persistent Sampling: Enhancing the Efficiency of Sequential Monte Carlo."

## Checklist

- [x] We should be able to understand what the PR does from its title only
- [x] There is a high-level description of the changes
- [x] There are links to all the relevant issues, discussions and PRs
- [x] The branch is rebased on the latest `main` commit
- [x] Commit messages follow [guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html)
- [x] The code respects the current naming conventions
- [x] Docstrings follow the [numpy style guide](https://numpydoc.readthedocs.io/en/latest/format.html)
- [x] `pre-commit` is installed and configured, and ran before opening the PR
- [x] There are tests covering the changes
- [x] The doc is up-to-date
- [x] Examples added/updated

## Discussion and Future Work

Implementation decisions can should be discussed/improved:

- The adaptive PS sampler does not check if the max iterations is exceeded and continues by overriding the last available state, which would break the algorithm. This can be checked as part of the inference loops (decision so far), or an internal mechanism could be implemented
- Arrays in the PSState are padded to allow for JIT-compatibility. This increases the memory footprint by a lot. Potentially, more a more elegant solution could be possible

Improvements:

- For long schedules, the algorithm might be adaptable to store the last $N$ iterations, rather than all iterations. This would improve memory-efficiency, but the update equations need to be corrected for this.

## Disclaimer

This is my first proper PR to an external codebase, so I'm sure I made some amateur mistakes throughout the process of creating this PR. Please let me know, if I can fix/improve anything!

### CODE REVIEW

File: blackjax/smc/persistent_sampling.py [hunk 1]
```diff
@@ -0,0 +1,803 @@
...
+            )[0]
+        )
+
+    @property
+    def num_particles(self) -> int:
+        """Number of particles."""
+        return self.persistent_log_likelihoods.shape[1]
+
+
+class PSInfo(NamedTuple):
```
Comment (junpenglao) [inline blackjax/smc/persistent_sampling.py]:
```suggestion
class PersistentStateInfo(NamedTuple):
```

**Comment by proteneer** [comment]
Awesome! So glad to it finally make its way into blackjax - can't wait to play with it in the near future.

File: blackjax/smc/persistent_sampling.py [hunk 2]
```diff
@@ -0,0 +1,803 @@
...
+    tempering_schedule: Array
+    iteration: int | Array
+
+    @property
+    def particles(self) -> ArrayLikeTree:
+        """Particles in current iteration."""
+        return jax.tree.map(lambda x: x[self.iteration], self.persistent_particles)
+
+    @property
+    def lmbda(self) -> float | Array:
```
[the remark this answers is not in the document]
Comment (junpenglao) [reply blackjax/smc/persistent_sampling.py]: I just update the parameter name in #801, could you update the branch to main and remove the alias?

[the remark this answers is not in the document]
Comment (ChrisBoettner) [reply blackjax/smc/persistent_sampling.py]: Perfect, thanks! Done and done.

**Review by junpenglao — APPROVED** [review]
